src/main/java/com/ngs/stash/externalhooks/hook/ExternalMergeCheckHook.java

```diff
@@ -0,0 +1,222 @@
+        Repository repo = pr.getToRef().getRepository();
+
+        // compat with Stash < 3.2.0
+        String repoPath = this.properties.getRepositoryDir(repo).getAbsolutePath();
+
+        Settings settings = context.getSettings();
+        List<String> exe = new LinkedList<String>();
+        exe.add(this.getExecutable(
+            settings.getString("exe"),
+            settings.getBoolean("safe_path", false)).getPath());
+
+        if (settings.getString("params") != null) {
+            for (String arg : settings.getString("params").split("\r\n")) {
+                exe.add(arg);
+            }
+        }
+
+        StashUser currentUser = authCtx.getCurrentUser();
+        ProcessBuilder pb = new ProcessBuilder(exe);
+
+        Map<String, String> env = pb.environment();
+        env.put("STASH_USER_NAME", currentUser.getName());
+        if (currentUser.getEmailAddress() != null) {
+            env.put("STASH_USER_EMAIL", currentUser.getEmailAddress());
+        } else {
+            log.error("Can't get user email address. getEmailAddress() call returns null");
+        }
+        env.put("STASH_REPO_NAME", repo.getName());
+
+        boolean isAdmin = permissions.hasRepositoryPermission(
+            currentUser, repo, Permission.REPO_ADMIN);
+        env.put("STASH_IS_ADMIN", String.valueOf(isAdmin));
+
+        RepositoryCloneLinksRequest.Builder cloneLinksRequestBuilder =
+            new RepositoryCloneLinksRequest.Builder();
+
+        cloneLinksRequestBuilder.repository(repo);
+
+        RepositoryCloneLinksRequest cloneLinksRequest =
+            cloneLinksRequestBuilder.build();
+
+        Set<NamedLink> cloneLinks = this.repoService.getCloneLinks(
+            cloneLinksRequest
+        );
+
+        for (NamedLink link : cloneLinks) {
+            env.put(
+                "STASH_REPO_CLONE_" + link.getName().toUpperCase(),
+                link.getHref()
+            );
+        }
+
+        env.put(
+            "STASH_BASE_URL",
+            this.properties.getBaseUrl().toString()
+        );
+
+        env.put("STASH_PROJECT_NAME", repo.getProject().getName());
+        env.put("STASH_PROJECT_KEY", repo.getProject().getKey());
+
+        String summaryMsg = "Merge request failed";
+
+        pb.directory(new File(repoPath));
+        pb.redirectErrorStream(true);
+        try {
+            Process process = pb.start();
+            InputStreamReader input = new InputStreamReader(
+                process.getInputStream(), "UTF-8");
+            OutputStream output = process.getOutputStream();
+
+            output.write((
+                          pr.getFromRef().getRepository().getSlug() + " " +
+                          pr.getFromRef().getLatestChangeset() + " " +
+                          pr.getToRef().getRepository().getSlug() + " " +
+                          pr.getToRef().getLatestChangeset() + "\n"
+                          ).getBytes("UTF-8"));
+            output.close();
```

Comment:
First of all: **Thanks a lot for starting this! It's very much needed!**

On a first look, the stdin is problematic.

For starters, git hooks expect 3, not 4, elements in this order, and with these semantics `PREV_SHA NEW_SHA REF`. See also existing code https://github.com/ngsru/atlassian-external-hooks/blob/master/src/main/java/com/ngs/stash/externalhooks/hook/ExternalPreReceiveHook.java#L126 .

Secondly, if the source branch is relevant - **I can't think of a reason** -, then it should be made available via other means e.g. as an environment variable `STASH_PULL_REQUEST_BRANCH`.

Thirdly, getLatestChangeset() returns quite a lot, I think, rather than just the commit. It only becomes the commit, because of the implicit type cast. `getLatestCommit()` should be enough. Also does `getRepository().getSlug()` actually return the ref (branch)? I think it should be `getToRef().getDisplayId()`

I cannot test my suggestion at the moment, but how about:

```
 +                          pr.getFromRef().getLatestCommit() + " " +
 +                          pr.getToRef().getLatestCommit() + "\n"
 +                          pr.getToRef().getDisplayId() + "\n"
```